src/graphics/shadow.odin

```diff
@@ -0,0 +1,351 @@
+            make([]Image, frame_count),
+            make([]vk.Framebuffer, frame_count),
+            render_pass,
+            frame_count,
+            extent,
+        }
+    shadow_init_mat(device, &shadows, mat_cache, descriptor_pool)
+    shadow_fill_buffer(device, &shadows, scene, light_data, frame_count)
+    shadow_init_images(device, &shadows, frame_count, extent3D)
+
+    return shadows
+}
+
+shadow_init_images :: proc(device: vk.Device, shadows: ^ShadowContext, frame_count: int, extent: vk.Extent3D) {
+    for i in 0..<frame_count {
+        shadows.images[i] = 
+            buffers_create_image(device, extent)
+        shadows.framebuffers[i] = 
+            shadow_create_framebuffer(device, 
+                shadows.render_pass, 
+                &shadows.images[i].view,
+                extent,
+            )
+    }
+
+    return
+}
+
+shadow_init_mat :: proc(device: vk.Device, shadows: ^ShadowContext, mat_cache: ^MaterialCache, descriptor_pool: vk.DescriptorPool){
+    shadow_effect := mats_create_shadow_effect(device, mat_cache, shadows.render_pass)
+    assert(shadow_effect != nil)
+    shadows.mat = mats_create(mat_cache, "default_shadow", device, descriptor_pool, shadow_effect)
+    assert(shadows.mat != nil)
+}
+
+shadow_fill_buffer :: proc(device: vk.Device, shadows: ^ShadowContext, scene: ^Scene, lights: ^LightData, frame_count: int) {
+    /*
+    reserve(&shadows.data, len(lights^.data))
+    for light in lights^.data {
+        shadow := Shadow {
+            la.matrix4_translate_f32(vec3(-light.direction.xyz)),
+            la.matrix4_perspective_f32(45, f32(500) / f32(500), 0.1, 100),
+        }
+        append(&shadows.data, shadow)
+    }
+    shadows_size := len(shadows.data) * size_of(Shadow)
+    // Duplicate for frames
+    for i in 0..<frame_count { buffers_write(shadows.buffer, raw_data(shadows.data), Shadow, i * shadows_size)}
+    */
+    
+    builders.bind_descriptor_set(device, 
+                                {shadows.buffer.handle, 0, size_of(Shadow)},
+                                .UNIFORM_BUFFER_DYNAMIC,
+                                shadows.mat.descriptors[0], 0)
+    builders.bind_descriptor_set(device, 
+                                {scene.object_ubos.handle, 0, size_of(mat4)},
+                                .UNIFORM_BUFFER_DYNAMIC,
+                                shadows.mat.descriptors[3], 0)
+
+
+    shadow := Shadow {
+        la.matrix4_look_at(
+            vec3{0, 2, 0},
+            vec3{0, 0, 0},
+            vec3{0, 0, -1},
+        ),
+        //la.matrix4_translate_f32(vec3(-lights^.data[0].direction.xyz)),
+        la.matrix_ortho3d_f32(-5, 5, -5, 5, 0.1, 5),
+        //la.matrix4_perspective_f32(45, f32(500) / f32(500), 0.1, 100),
+    }
+    for i in 0..<frame_count { buffers_write(shadows.buffer, &shadow, Shadow, i)}
+}
+
+shadow_destroy :: proc(device: vk.Device, shadows: ^ShadowContext){
+    for i in 0..<shadows.frame_count {
```

Comment:
you need to destroy the slices that you make in your init functions (frame buffers, image views, images)